cli: TestLossOfQuorumRecovery failed

cli.TestLossOfQuorumRecovery [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9369590?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9369590?buildTab=artifacts#/) on release-23.1 @ [d3b6cded68ee6227c0532f45256c379838329acd](https://github.com/cockroachdb/cockroach/commits/d3b6cded68ee6227c0532f45256c379838329acd):


```
id	is_live	replicas	is_decommissioning	membership	is_draining
2	true	63	true	decommissioning	false
3	true	63	true	decommissioning	false
    debug_recover_loss_of_quorum_test.go:278: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:278
        	Error:      	Received unexpected error:
        	            	unable to retrieve conf reader: span configs not available
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.forceScanAndProcess
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/queue_helpers_testutil.go:34
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).ForceConsistencyQueueProcess
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/queue_helpers_testutil.go:99
        	            	  | github.com/cockroachdb/cockroach/pkg/cli.TestLossOfQuorumRecovery.func2
        	            	  | 	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:279
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:150
        	            	  | github.com/cockroachdb/cockroach/pkg/util/syncutil.(*IntMap).Range
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/syncutil/int_map.go:352
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).VisitStores
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:149
        	            	  | github.com/cockroachdb/cockroach/pkg/cli.TestLossOfQuorumRecovery
        	            	  | 	github.com/cockroachdb/cockroach/pkg/cli/debug_recover_loss_of_quorum_test.go:278
        	            	  | [...repeated from below...]
        	            	Wraps: (2) unable to retrieve conf reader
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_gossip.go:159
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (4) span configs not available
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
        	Test:       	TestLossOfQuorumRecovery
        	Messages:   	Failed to force replicas to consistency queue
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0ede0149c79dc299cd458e6d0ea4f601/logTestLossOfQuorumRecovery386017736
--- FAIL: TestLossOfQuorumRecovery (38.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLossOfQuorumRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26384

--- a/pkg/cli/debug_recover_loss_of_quorum_test.go
+++ b/pkg/cli/debug_recover_loss_of_quorum_test.go
@@ -257,7 +257,8 @@ func TestLossOfQuorumRecovery(t *testing.T) {
 	// attempt. That would increase number of replicas on system ranges to 5 and we
 	// would not be able to upreplicate properly. So we need to decommission old nodes
 	// first before proceeding.
-	grpcConn, err := tcAfter.Server(0).RPCContext().GRPCDialNode(tcAfter.Server(0).ServingRPCAddr(), tcAfter.Server(0).NodeID(), rpc.DefaultClass).Connect(ctx)
+	grpcConn, err := tcAfter.Server(0).RPCContext().GRPCDialNode(
+		tcAfter.Server(0).ServingRPCAddr(), tcAfter.Server(0).NodeID(), rpc.DefaultClass).Connect(ctx)
 	require.NoError(t, err, "Failed to create test cluster after recovery")
 	adminClient := serverpb.NewAdminClient(grpcConn)
 
@@ -272,6 +273,8 @@ func TestLossOfQuorumRecovery(t *testing.T) {
 			return nil
 		}), "Failed to activate replication queue")
 	}
+	require.NoError(t, tcAfter.WaitForZoneConfigPropagation(),
+		"Failed to ensure zone configs are propagated")
 	require.NoError(t, tcAfter.WaitForFullReplication(), "Failed to perform full replication")
 
 	for i := 0; i < len(tcAfter.Servers); i++ {

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -1446,19 +1446,30 @@ func (tc *TestCluster) WaitFor5NodeReplication() error {
 		// conformance since zone configs are propagated synchronously.
 		// Generous timeout is added to allow rangefeeds to catch up. On startup
 		// they could get delayed making test to fail.
-		now := tc.Server(0).Clock().Now()
-		for _, s := range tc.Servers {
-			scs := s.SpanConfigKVSubscriber().(spanconfig.KVSubscriber)
-			if err := testutils.SucceedsSoonError(func() error {
-				if scs.LastUpdated().Less(now) {
-					return errors.New("zone configs not propagated")
-				}
-				return nil
-			}); err != nil {
-				return err
-			}
+		err := tc.WaitForZoneConfigPropagation()
+		if err != nil {
+			return err
 		}
 		return tc.WaitForFullReplication()
+	}
+	return nil
+}
+
+// WaitForZoneConfigPropagation ensures that all span config subscribers caught
+// up till now. That would guarantee that zone configs set prior to this call
+// are applied.
+func (tc *TestCluster) WaitForZoneConfigPropagation() error {
+	now := tc.Server(0).Clock().Now()
+	for _, s := range tc.Servers {
+		scs := s.SpanConfigKVSubscriber().(spanconfig.KVSubscriber)
+		if err := testutils.SucceedsSoonError(func() error {
+			if scs.LastUpdated().Less(now) {
+				return errors.New("zone configs not propagated")
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
 	}
 	return nil
 }

